### ISSUE TITLE
[feature request] Different interpolation algos for 'grid_sample' function

### ISSUE THREAD

**Issue by alasin** [pytorch/pytorch#5565]
The 2D grid interpolator function 'grid_sample' only does bilinear sampling from the source image at the moment. It would be useful to have 'nearest' and other samplings available. The function call does take 'mode' as a parameter but only bilinear sampling seems to have been implemented. It would especially be useful if we have to use 'grid_sample' for class label maps for segmentation, and the grid (source indices) is a Variable.

**Comment by ailzhang** [comment]
This is not with grid_sampler but might be useful in your case? https://pytorch.org/docs/master/nn.html?highlight=interpolate#torch.nn.functional.interpolate.

**Comment by fmassa** [comment]
I also agree that having multiple interpolation modes in `grid_sample` could be desirable in some applications.

**Comment by creotiv** [comment]
Trilinear interpolation should be added, cause in case of 5D input using bilinear interpolation is wrong cause we have 3 position coordinates

**Comment by ezyang** [comment]
We'll accept a patch that adds trilinear interpolation.

**Comment by ksnzh** [comment]
Hope trilinear interpolation will be added.

**Comment by chinibinibin** [comment]
What is the grid_sample method doing at all in the 5D case, e.g. with 3D interpolation data, when 'bilinear' mode is active?

**Comment by creotiv** [comment]
@chinibinibin by my observation it just not use third shape dimension. So result in terms of image becoming pixelated.

**Comment by chinibinibin** [comment]
Well this would be critical, as it is not clear on which slice it does the interpolation then.
Looking empirically at results it seems like it is already doing some 3D interpolation currently... not sure.
I need to check this.

**Comment by christopher-beckham** [comment]
Does anyone have an implementation of this? Even if it's written in PyTorch (rather than C) code?

**Comment by creotiv** [comment]
@chinibinibin no, it's doing only 2d interpolation. i tested it on HDRNet network from google, and it was giving wrong results. Thus network doesnt work. So problem still exists. Especialy when we have 3D space but doesnt have interpolation for it :(

**Comment by christopher-beckham** [comment]
I'm no expert on the algorithm for trilinear/bilinear sampling, but it appears that the z coordinates are being used still in the bilinear case, as seen here: https://github.com/pytorch/pytorch/blob/81bf73643b6552a63794fb889238aaf0c2a7baa6/aten/src/ATen/native/GridSampler.cpp#L68-L147

??

**Comment by VolodymyrAhafonov** [comment]
Hey guys, any updates about trilinear grid sample?

**Comment by christopher-beckham** [comment]
@VolodymyrAhafonov @creotiv a while back I had ported some Numpy code I found on the net to do the trilinear sampling, though I don't know if it's correct, so no guarantees. It's also quite slow (because it's not a fast C/CUDA implementation).

The way it's called in `stn` is a bit hacky and unclean but it's what I had to do at the time to get it to work.

```
def interpolate_trilinear(self, img, x, y, z):

    # ...

    x0 = torch.floor(x).long()
    x1 = x0 + 1
    y0 = torch.floor(y).long()
    y1 = y0 + 1
    z0 = torch.floor(z).long()
    z1 = z0 + 1

    ix = torch.zeros_like(x0)
    for j in range(x.size(0)):
        ix[j] += j

    x0 = torch.clamp(x0, min=0, max=img.shape[2] - 1)
    x1 = torch.clamp(x1, min=0, max=img.shape[2] - 1)
    y0 = torch.clamp(y0, min=0, max=img.shape[3] - 1)
    y1 = torch.clamp(y1, min=0, max=img.shape[3] - 1)
    z0 = torch.clamp(z0, min=0, max=img.shape[4] - 1)
    z1 = torch.clamp(z1, min=0, max=img.shape[4] - 1)

    x_ = x - x0.float()
    y_ = y - y0.float()
    z_ = z - z0.float()

    out = (img[:, ix, x0, y0, z0]*(1-x_)*(1-y_)*(1-z_) +
           img[:, ix, x1, y0, z0]*x_*(1-y_)*(1-z_) +
           img[:, ix, x0, y1, z0]*(1-x_)*y_*(1-z_) +
           img[:, ix, x0, y0, z1]*(1-x_)*(1-y_)*z_ +
           img[:, ix, x1, y0, z1]*x_*(1-y_)*z_ +
           img[:, ix, x0, y1, z1]*(1-x_)*y_*z_ +
           img[:, ix, x1, y1, z0]*x_*y_*(1-z_) +
           img[:, ix, x1, y1, z1]*x_*y_*z_)

    return out

def stn(self, x, theta):

    grid = F.affine_grid(theta, x.size())
    if x.is_cuda:
        grid = grid.to(self.device)

    if self.interp_mode == 'bilinear':
        out = F.grid_sample(x, grid, padding_mode='zeros')
    elif self.interp_mode == 'trilinear':
        im_sz = x.size(-1)
        out = self.interpolate_trilinear(x.transpose(0,1),
                                         (grid[:,:,:,:,2]*0.5 + 0.5)*im_sz,
                                         (grid[:,:,:,:,1]*0.5 + 0.5)*im_sz,
                                         (grid[:,:,:,:,0]*0.5 + 0.5)*im_sz)
        out = out.transpose(1, 0).contiguous()
    else:
        raise Exception("this interp mode not implemented")

    return out

```

**Comment by tedyhabtegebrial** [comment]
here is a python implementation of trilinear grid sampling;  it might not be optimized for performance.
https://github.com/tedyhabtegebrial/PyTorch-Trilinear-Interpolation

**Comment by aTamaz** [comment]
@ezyang 

In the current version (1.5.0 on Ubuntu 16.04, Python 3.5), the function [grid_sample](https://pytorch.org/docs/stable/nn.functional.html#grid-sample) actually performs as expected, i.e. using trilinear interpolation for 5D case -- both on CPU and CUDA. The nearest interpolation option also works fine. Haven't tested previous versions.

There's one caveat though -- if your input has the dimensions NCDHW, and the grid's three coordinates that are provided for each point (in the last dimension) are XYZ, where X corresponds to image's width, Y to height, and Z to depth.

Here's an example that I've used for testing:

```
import torch
import torch.nn.functional as fn

# N x C x D x H x W image
image = torch.arange(0, 8, dtype=torch.float32).reshape([1, 1, 2, 2, 2])

# sampling positions with (W, H, D) coordinates between 0 and 1
w_coord = 1/5
h_coord = 3/4
d_coord = 2/3
sampling_position = torch.tensor([w_coord, h_coord, d_coord], dtype=torch.float32)
# bring the sampling coordinates into the range -1..1
sampling_position = 2 * sampling_position - 1
# add the N, D, H, W dimensions
sampling_position = torch.reshape(sampling_position, [1, 1, 1, 1, *sampling_position.shape])

sample = fn.grid_sample(image, sampling_position, 'bilinear', align_corners=True)
print(sample)

# Manual computation
w_00 = image[0, 0, 0, 0, 0] * (1 - w_coord) + image[0, 0, 0, 0, 1] * w_coord
w_01 = image[0, 0, 0, 1, 0] * (1 - w_coord) + image[0, 0, 0, 1, 1] * w_coord
w_10 = image[0, 0, 1, 0, 0] * (1 - w_coord) + image[0, 0, 1, 0, 1] * w_coord
w_11 = image[0, 0, 1, 1, 0] * (1 - w_coord) + image[0, 0, 1, 1, 1] * w_coord

w_0 = w_00 * (1 - h_coord) + w_01 * h_coord
w_1 = w_10 * (1 - h_coord) + w_11 * h_coord

w = w_0 * (1 - d_coord) + w_1 * d_coord

print(w)
```

So it'd be nice if the documentation could be updated -- to reflect the relation of XYZ sampling coordinates to WHD of the input, as well as explicitly mention trilinear sampling in 5D case. Otherwise the code works as expected.

**Comment by creotiv** [comment]
Tri-linear interpolation

![](https://cdn.mathpix.com/snip/images/3UUbvcO6OoA7TYxUvmUNsUdHOGXBm2M9r33GdDYC7t4.original.fullsize.png)
tau = max(1-| * |,0) - interpolation kernel
sx,sy,d - width,height, z of input
 i,j,k- current input position
x,y,g[x,y] - x,y,value of grid

And i think grid_sample not doing this, and that is a problem

**Comment by aTamaz** [comment]
@creotiv my example above shows that grid_sample correctly implements trilinear interpolation as defined e.g. on the [wiki](https://en.wikipedia.org/wiki/Trilinear_interpolation#Method).

**Comment by KrakenLeaf** [comment]
Hello. Following the discussion above, I wanted to ask whether it is possible to add additional interpolation kernels to grid_sample?
I mostly refer to truncated cosine / truncated sinc interpolation kernels or cubic interpolation for 5D data.
The reason is image blur after interpolation (5D with trilinear) - I'm working on medical images (i.e. MR), and the resulting resampled image has some blur, compared with the MR image I started from, which happens since a linear interpolation kernel is less suited than e.g. a tapered sinc kernel for these images. Thanks.
